lib/node_modules/@stdlib/repl/lib/multiline_handler.js

```diff
@@ -0,0 +1,348 @@
+*
+* @private
+* @name _moveUp
+* @memberof MultilineHandler.prototype
+* @type {Function}
+* @returns {void}
+*/
+setNonEnumerableReadOnly( MultilineHandler.prototype, '_moveUp', function moveUp() {
+	var x;
+
+	// If already at the first line, ignore...
+	if ( this._multiline.line <= 0 ) {
+		return;
+	}
+	this._repl._cmd[ this._multiline.line ] = this._rli.line; // update current line in command
+
+	// Make sure the cursor never exceeds the length of the line:
+	x = min( this._rli.cursor, this._repl._cmd[ this._multiline.line - 1 ].length ); // eslint-disable-line max-len
+	this._moveCursor( x, -1 );
+});
+
+/**
+* Moves cursor down to the next line.
+*
+* @private
+* @name _moveDown
+* @memberof MultilineHandler.prototype
+* @type {Function}
+* @returns {void}
+*/
+setNonEnumerableReadOnly( MultilineHandler.prototype, '_moveDown', function moveDown() {
+	var x;
+
+	// If already at the last line, ignore...
+	if ( this._multiline.line >= this._multiline.lines.length - 1 ) {
+		return;
+	}
+	this._repl._cmd[ this._multiline.line ] = this._rli.line; // update current line in command
+
+	// Make sure the cursor never exceeds the length of the line:
+	x = min( this._rli.cursor, this._repl._cmd[ this._multiline.line + 1 ].length ); // eslint-disable-line max-len
+	this._moveCursor( x, 1 );
+});
+
+/**
+* Moves cursor left to the end of the previous line.
+*
+* @private
+* @name _moveLeft
+* @memberof MultilineHandler.prototype
+* @type {Function}
+* @returns {void}
+*/
+setNonEnumerableReadOnly( MultilineHandler.prototype, '_moveLeft', function moveLeft() {
+	// If already at the first line, ignore...
+	if ( this._multiline.line <= 0 ) {
+		return;
+	}
+	this._repl._cmd[ this._multiline.line ] = this._rli.line; // update current line in command
+
+	// Move cursor to the end of the previous line:
+	this._moveCursor( this._repl._cmd[ this._multiline.line - 1 ].length, -1 );
+});
+
+/**
+* Moves cursor right to the start of the next line.
+*
+* @private
+* @name _moveRight
+* @memberof MultilineHandler.prototype
+* @type {Function}
+* @returns {void}
+*/
+setNonEnumerableReadOnly( MultilineHandler.prototype, '_moveRight', function moveRight() {
+	// If already at the last line, ignore...
+	if ( this._multiline.line >= this._multiline.lines.length - 1 ) {
+		return;
+	}
+	this._repl._cmd[ this._multiline.line ] = this._rli.line; // update current line in command
+
+	// Move cursor to the start of the next line:
+	this._moveCursor( 0, 1 );
+});
+
+/**
+* Simulates a backspace by removing the current line and appending it to the previous line.
+*
+* @private
+* @name _backspace
+* @memberof MultilineHandler.prototype
+* @type {Function}
+* @returns {void}
+*/
+setNonEnumerableReadOnly( MultilineHandler.prototype, '_backspace', function backspace() {
+	var x;
+
+	// If already at the first line, ignore...
+	if ( this._multiline.line <= 0 ) {
+		return;
+	}
+	// Save cursor position:
+	x = this._repl._cmd[ this._multiline.line - 1 ].length;
```

Comment:
Accessing private properties on the REPL instance does not seem ideal, as that makes implicit coupling. The same goes for `_inputPrompt`.

Reply:
I cached a reference to these REPL properties inside the `multilineHandler` class, is that okay?

Reply:
Yeah, although this practice is still not ideal (aside: I know it is common elsewhere in the implementation for code that I originally wrote), as we're effectively sharing state. At some point, we should refactor things to be cleaner, perhaps by following a design pattern from _Design Patterns_.